pageserver/src/repository.rs

```diff
@@ -1,24 +1,167 @@
-use crate::relish::*;
-use crate::walrecord::MultiXactMember;
+use crate::keyspace::KeyPartitioning;
+use crate::walrecord::ZenithWalRecord;
 use crate::CheckpointConfig;
-use anyhow::Result;
+use anyhow::{bail, Result};
 use bytes::Bytes;
-use postgres_ffi::{MultiXactId, MultiXactOffset, TransactionId};
 use serde::{Deserialize, Serialize};
-use std::collections::HashSet;
-use std::ops::{AddAssign, Deref};
+use std::fmt;
+use std::ops::{AddAssign, Range};
 use std::sync::{Arc, RwLockReadGuard};
 use std::time::Duration;
 use zenith_utils::lsn::{Lsn, RecordLsn};
 use zenith_utils::zid::ZTimelineId;
 
-/// Block number within a relish. This matches PostgreSQL's BlockNumber type.
-pub type BlockNumber = u32;
+#[derive(Debug, Clone, Copy, Hash, PartialEq, Eq, Ord, PartialOrd, Serialize, Deserialize)]
+/// Key used in the Repository kv-store.
+///
+/// The Repository treates this as an opaque struct, but see the code in pgdatadir_mapping.rs
+/// for what we actually store in these fields.
+pub struct Key {
+    pub field1: u8,
+    pub field2: u32,
+    pub field3: u32,
+    pub field4: u32,
+    pub field5: u8,
+    pub field6: u32,
+}
```

Comment:
This struct is growing on me.

We can rewrite it like this for readability if needed:
```
pub struct Key {
    pub noncontiguous_component: [u8; 14],
    pub contiguous_component: [u8; 4],
}
```
Might be a premature refactor, let me read the rest of the PR :)

Reply:
Yeah, this could be done in many different ways. At first I tried to cram it into a u128, but that's not quite big enough. could be `field1: u128, field2: u16`. But one nice property of this layout is that it matches the RelFileNode fields, so that when you look at a Key in a debugger, you can easily see the relation's 'relnode', forknum and block number in separate fields, in field4-field6.

Another thought is that since we have the "RelDirectory" metadata that list all the relations anyway, we could assign an arbitrary ID to each relation that's separate from its RelFileNode. So when you create a relation, assign it a new 64-bit ID, and store that in the RelDirectory. That would allow shortening the Key, which would speed up processing all over the layer code. I think it could be 96 bits: u64 for the relation ID and u32 for block number. The downside is that get_page() call would need to first look up the relation's ID in the metadata. You could cache it, but then you'd need to maintain the cache, so it would add some complexity. And then you couldn't immediately decipher the relfilenode from the Key when you're debugging.